dags/destinations/cm360oci.py

```diff
@@ -0,0 +1,288 @@
+
+  def send_data(self, input_data: List[Mapping[str, Any]], dry_run:bool) -> Optional[RunResult]:
+    """Builds payload and sends data to CM360 API."""
+
+    valid_conversions = []
+    invalid_conversions = []
+    encryption_info = {}
+
+    for i,entry in enumerate(input_data):
+      conversion = {}
+      for conversion_field in CM_CONVERSION_FIELDS:
+        if conversion_field == "timestamp_micros":
+          timestamp_micros = self._parse_timestamp_micros(entry)
+          if timestamp_micros:
+            conversion[conversion_field] = timestamp_micros
+          else:
+            conversion[conversion_field] = ""
+        elif conversion_field == "conversion_kind":
+          conversion["kind"] = str(entry.get(conversion_field, ""))
+        else:
+          conversion[conversion_field] = str(entry.get(conversion_field, ""))
+      
+      if self.validate_conversion(conversion):
+        valid_conversions.append(conversion)
+      else:
+        invalid_conversions.append((i, errors.ErrorNameIDMap.CM_HOOK_ERROR_INVALID_CONVERSION_EVENT))
+
+    if valid_conversions:
+      payload = {}
+      payload["encryptionInfo"] = self.encryption_info
+      payload["conversions"] = valid_conversions
+
+      if not dry_run:
+        try:
+          self._send_payload(payload)
+        except (
+              errors.DataOutConnectorSendUnsuccessfulError,
+          ) as error:
+            send_error = error.error_num
+      else:
+        print(
+          "Dry-Run: CM conversions event will not be sent."
+        )
+
+    run_result = RunResult(
+      successful_hits=len(valid_conversions),
+      failed_hits=len(invalid_conversions),
+      error_messages=[str(error[1]) for error in invalid_conversions],
+      dry_run=dry_run,
+    )
+
+    return run_result
+
+  @staticmethod
+  def schema() -> Optional[ProtocolSchema]:
+    return ProtocolSchema(
+      "CM360OCI",
+      [
+        ("profile_id", str, Field(description="A Camaign Manager Profile ID .")),
+        ("encryptionEntityType", str, Field(description="The encryption entity type.")),
+        ("encryptionEntityId", str, Field(description="The encryption entity ID.")),
+        ("encryptionSource", str, Field(description="Describes whether the encrypted cookie was received from ad serving or from Data Transfer.")),
+        ("kind", str, Field(description="Identifies what kind of resource this is.")),
+        ("access_token", str, Field(description="A Campaign Manager 360 access token.")),
+        ("refresh_token", str, Field(description="A Campaign Manager 360 API refresh token.")),
+        ("token_uri", str, Field(description="A Campaign Manager 360 API token uri.")),
+        ("client_id", str, Field(description="An OAuth2.0 Web Client ID.")),
+        ("client_secret", str, Field(description="An OAuth2.0 Web Client Secret.")),
+      ]
+    )
+
+  def fields(self) -> Sequence[str]:
+    return CM_CONVERSION_FIELDS
+
+  def batch_size(self) -> int:
+    return 10000
```

Comment:
I think this should be 1000 according to https://developers.google.com/doubleclick-advertisers/guides/conversions_faq?

Reply:
fixed